부산대 BE_오선재 - Step2 회원 로그인

5seonjae opened the pull request:
### 변경 사항
- **Member 엔티티 / DTO**  
  - `MemberRegisterRequestDto`, `MemberRegisterResponseDto` 추가  
- **Repository (JdbcClient 기반)**  
  - `MemberJdbcRepository` 구현: `findByEmail`, `existsByEmail`, `save` 메서드 추가  
- **Service**  
  - `AuthServiceJdbc` (`loginMember`, `registerMember`) 구현  
    - 이메일 중복 검사, 비밀번호 암호화/검증  
    - 로그인 시 Basic 인증 헤더 디코딩 후 JWT 발급  
- **Controller**  
  - `/api/members/register` (회원가입) 및 `/api/members/login` (로그인) 엔드포인트 구현  
  - 커스텀 예외(`MissingAuthorizationHeaderException`, `InvalidAuthorizationHeaderException`, `InvalidCredentialsException`)로 인증 흐름 분기  
- **Global Exception Handler**  
  - `@ControllerAdvice`에서 예외별 상태 코드(401/403) 및 `{ "error": ... }` 바디 반환  
- **테스트**  
  - MockMvc를 이용한 회원가입·로그인 성공 및 예외 시나리오 테스트 전부 추가  
    - 빈 헤더, 잘못된 형식, 자격 증명 오류 등 커버  

### 특징
- Spring Security 프레임워크를 도입하지 않고, 순수 JDBC(`JdbcClient`)와 커스텀 예외/필터로 인증 로직을 직접 구현  
- JWT 발급 및 검증은 `TokenService`를 통해 담당  

### 보안 설정
- 민감 정보(jwt.secret, jwt.expiration-ms)를 `.env` 파일로 분리하여 관리
  - application.properties에 `spring.config.import=file:.env[.properties]` 설정 추가

### 실행 예시

- H2 콘솔 접속
  - ![image](https://github.com/user-attachments/assets/ae033b04-5985-42b4-9f49-a28275079a94)

- 회원가입 성공
  - ![image](https://github.com/user-attachments/assets/1f474e8c-b2be-4d89-aa72-0fbcf8fa2245)
  - ![image](https://github.com/user-attachments/assets/05e6b76b-50a6-4662-b29d-a98756515c36)

- 회원가입 실패 (이메일 중복)
  - ![image](https://github.com/user-attachments/assets/bb26debd-28f9-4df3-907e-6c7006c68304)

- 회원가입 실패 (이메일 형식 오류)
  - ![image](https://github.com/user-attachments/assets/16a6727b-dc4a-408c-a748-23423253237e)

- 회원가입 실패 (6자 미만 비밀번호)
  - ![image](https://github.com/user-attachments/assets/1d46502e-32cf-4d22-8b6c-c9f6ea418617)

- 로그인 성공
  - ![image](https://github.com/user-attachments/assets/518c2262-4581-42f9-a08f-98049ef07862)

- 로그인 실패 (헤더 누락)
  - ![image](https://github.com/user-attachments/assets/2b2f3396-f00a-4166-9abb-47aa7279897d)

- 로그인 실패 (잘못된 프리픽스)
  - ![image](https://github.com/user-attachments/assets/b2bb76d9-a92f-4326-9935-b9a6d57a7e33)

- 로그인 실패 (디코딩 오류/구분자 누락)
  - ![image](https://github.com/user-attachments/assets/daf2e95c-84b0-43a6-a0c4-bb95dc31fd6d)

- 로그인 실패 (잘못된 비밀번호)
  - ![image](https://github.com/user-attachments/assets/e7c2031c-9599-4254-809f-91640d68b1bb)

### 미배포(선택) 사항
- 다음 단계에서 Spring Security 및 JWT 필터 적용, 보호된 API 엔드포인트 구현 예정

5seonjae committed:
docs(auth): README에 회원 로그인 기능 및 API 명세 추가

5seonjae committed:
feat: Member 엔티티 추가 및 유효성 검증 로직 구현

- `Member` 클래스 생성
- 생성자에서 이메일·비밀번호 유효성 검증 수행
    - 이메일 null/빈값 검사 및 `. +@.+\..+` 정규식 기반 형식 검사
    - 비밀번호 null/빈값 검사 및 최소 6자 길이 검사
- 검증 실패 시 `IllegalArgumentException`을 통해 예외 발생

5seonjae committed:
feat: 회원 가입용 MemberRegisterRequestDto 추가 및 Bean Validation 적용

- `MemberRegisterRequestDto` 클래스 생성
- `email` 필드에 `@NotBlank`, `@Email` 애너테이션 추가하여 필수 입력 및 형식 검증
- `password` 필드에 `@NotBlank`, `@Size(min=6)` 애너테이션 추가하여 필수 입력 및 최소 길이 검증
- 생성자 및 Getter 메서드 구현

5seonjae committed:
feat: 회원가입 응답용 MemberRegisterResponseDto 추가

- `MemberRegisterResponseDto` 클래스 생성
- JWT 액세스 토큰을 담는 `token` 필드 선언
- 생성자를 통해 `token` 필드 초기화
- 클라이언트 응답을 위한 `getToken()` getter 메서드 구현

5seonjae committed:
feat: MemberRepository 클래스 생성 및 JdbcClient 의존성 주입 구현

- `gift.repository.MemberRepository` 클래스 추가
- 생성자를 통해 `JdbcClient` 인스턴스 주입받도록 설정

5seonjae committed:
refactor: Member 생성자 오버로드 및 ID 처리 로직 개선

- `Member(Long id, String email, String password)` 생성자 추가
- 기존 `Member(String email, String password)` 생성자를 새 생성자 호출 형태로 위임
- 등록 시 ID 없이도 생성 가능하도록 `this(null, email, password)` 형태로 설정

5seonjae committed:
feat: MemberRepository에 회원 저장(save) 기능 구현

- `save(Member)` 메서드 추가
- SQL `INSERT INTO member (email, password)` 쿼리 실행
- `GeneratedKeyHolder` 이용해 자동 생성된 `id` 조회 및 설정
- 저장된 정보를 포함해 새로운 `Member` 객체 반환

5seonjae committed:
feat: MemberService 클래스 추가 및 MemberRepository 의존성 주입 설정

- `gift.service.MemberService` 클래스 생성
- 생성자를 통해 `MemberRepository` 주입받도록 구현

5seonjae committed:
feat: MemberService에 회원 등록 기능(registerMember) 메서드 추가

- `registerMember(Member)` 메서드 구현
- 내부에서 `MemberRepository.save(member)` 호출하여 회원 저장 로직 위임

5seonjae committed:
feat: MemberController 클래스 추가 및 기본 URL 매핑 설정

- `@RestController` 및 `@RequestMapping("/api/members")` 애노테이션 적용
- `MemberService` 주입을 위한 생성자 구현

5seonjae committed:
feat: MemberController에 회원 생성 API(createMember) 엔드포인트 추가

- `@PostMapping` 메서드 `createMember` 구현
- `@RequestBody @Valid MemberRegisterRequestDto`로 요청 DTO 검증
- `MemberService.registerMember` 호출 후 저장된 `Member` 반환
- `ResponseEntity`로 HTTP 201 Created 상태코드 설정

5seonjae committed:
feat: schema.sql에 members 테이블 생성 DDL 추가

- `members` 테이블 정의 추가
    - `id` BIGINT PRIMARY KEY AUTO_INCREMENT
    - `email` VARCHAR(255) NOT NULL
    - `password` VARCHAR(255) NOT NULL

5seonjae committed:
fix: save 메서드 SQL에서 테이블 이름 오타 수정

- `INSERT INTO member` → `INSERT INTO members` 로 변경

5seonjae committed:
chore: .gitignore에 .env 파일 무시 설정 추가

- 프로젝트 루트에 .env를 무시하도록 .gitignore에 등록

5seonjae committed:
chore: JJWT 라이브러리 의존성 추가

- `jjwt-api`, `jjwt-impl`, `jjwt-jackson` 버전 0.12.6 추가

5seonjae committed:
application.properties에 JWT 관련 설정 추가 및 H2 DB 설정 유지

- `jwt.secret`에 JWT 토큰 서명용 시크릿 키(Base64 인코딩) 추가
- `jwt.expiration-ms`에 토큰 만료 시간을 3,600,000ms(1시간)로 설정
- 기존 Spring 애플리케이션 이름(`spring.application.name`) 및 H2 메모리 DB 설정(`spring.datasource.*`, `spring.sql.init.mode`, H2 콘솔 설정) 그대로 유지

5seonjae committed:
feat: TokenService 구현 및 JWT 설정 프로퍼티 주입

- `TokenService` 클래스 생성
- `@Value("${jwt.secret}")`로 서명용 시크릿 키 주입
- `@Value("${jwt.expiration-ms}")`로 토큰 만료 시간 주입
- `Keys.hmacShaKeyFor(secret.getBytes())`를 통해 JWT 서명용 `Key` 생성

5seonjae committed:
feat: 회원가입 흐름에 JWT 토큰 생성 및 API 응답 DTO 적용

- `TokenService.generateToken(Long memberId, String email)` 메서드 추가
    - Subject에 `memberId`, custom claim에 `email` 포함
    - 발행 시간(`iat`) 및 만료 시간(`exp`; `jwt.expiration-ms`) 설정
    - HS256 알고리즘으로 HMAC SHA 키(`jwt.secret`) 서명
- `MemberService.registerMember(MemberRegisterRequestDto)` 반환 타입을 `String`으로 변경
    - 회원 저장 후 `TokenService.generateToken` 호출하여 JWT 토큰 생성 및 반환
- `MemberController.createMember()` API 응답을 `MemberRegisterResponseDto`로 변경
    - 서비스가 반환한 토큰을 DTO에 담아 HTTP 201로 응답

5seonjae committed:
feat: application.properties에 외부 환경 파일 import 설정 추가

- `spring.config.import=file:.env[.properties]` 추가하여 .env(.properties) 파일에서 환경 변수 로드

5seonjae committed:
feat: GlobalExceptionHandler에 중복 이메일 예외 처리 핸들러 추가

- `DuplicateKeyException` 발생 시 HTTP 409 CONFLICT 응답
- 예외 메시지를 `Map<String, String>` 형태의 JSON 바디로 반환

5seonjae committed:
feat: MemberRepository에 이메일 중복 확인용 existByEmail 메서드 추가

- `SELECT COUNT(*) FROM members WHERE email = ?` 쿼리로 이메일 존재 여부 조회
- `jdbcClient.query(Integer.class).optional().orElse(0)` 사용해 결과를 `Integer count`로 매핑
- `count > 0`을 반환하여 중복 여부를 boolean으로 제공

5seonjae committed:
feat: 회원가입 시 이메일 중복 검사 및 예외 처리 추가

- `MemberService.registerMember` 메서드에 `memberRepository.existByEmail(requestDto.getEmail())` 호출해 중복 확인
- 중복일 경우 `DuplicateKeyException("이미 사용 중인 이메일입니다.")` 예외 발생

5seonjae committed:
refactor: 회원가입 API 경로를 `/register`로 변경

- `MemberController.createMember`의 `@PostMapping` 경로를 `/register`로 수정

5seonjae committed:
feat: MemberRepository에 이메일 기반 조회 메서드 추가

- `mapRowToMember(ResultSet, int)` 메서드로 `ResultSet` → `Member` 매핑 구현
- `findByEmail(String email)` 메서드 추가하여 이메일로 회원 조회 기능 제공 (`Optional<Member>` 반환)

5seonjae committed:
feat(auth): 로그인 처리 로직 구현 및 예외 처리

- `loginMember(String email, String password)` 메서드 추가
    - `findByEmail(email)`으로 회원 조회 후 없으면 `IllegalArgumentException("로그인 정보가 올바르지 않습니다.")` 발생
    - 조회된 회원의 비밀번호 불일치 시 동일한 예외 발생
    - 인증에 성공하면 `tokenService.generateToken(member.getId(), member.getEmail())`을 호출해 JWT 반환

5seonjae committed:
feat: InvalidAuthorizationHeaderException 커스텀 예외 클래스 추가

- `InvalidAuthorizationHeaderException` 클래스 생성
- `RuntimeException`을 상속하고, 메시지 전달용 생성자 구현

5seonjae committed:
feat: InvalidCredentialsException 커스텀 예외 클래스 추가

- 인증 실패 시 사용되는 `InvalidCredentialsException` 클래스 생성
- `RuntimeException`을 상속하고, 메시지 전달용 생성자 구현

5seonjae committed:
feat: MissingAuthorizationHeaderException 커스텀 예외 클래스 추가

- Authorization 헤더 누락 시 발생시키는 `MissingAuthorizationHeaderException` 클래스 생성
- `RuntimeException`을 상속하고, 메시지 전달용 생성자 구현

5seonjae committed:
feat: GlobalExceptionHandler에 인증 실패 예외 처리 핸들러 추가

- `InvalidCredentialsException` 발생 시 HTTP 403 FORBIDDEN 응답
- 예외 메시지를 `Map<String, String>` 형태의 JSON 바디로 반환

5seonjae committed:
feat: GlobalExceptionHandler에 Authorization 헤더 누락 예외 처리 핸들러 추가

- `MissingAuthorizationHeaderException` 발생 시 처리
- HTTP 401 UNAUTHORIZED 응답
- 예외 메시지를 `Map<String, String>` 형태의 JSON 바디로 반환

5seonjae committed:
feat: GlobalExceptionHandler에 Authorization 헤더 형식 오류 예외 처리 핸들러 추가

- `InvalidAuthorizationHeaderException` 발생 시 처리
- HTTP 401 UNAUTHORIZED 응답
- 예외 메시지를 `Map<String, String>` 형태의 JSON 바디로 반환

5seonjae committed:
feat: 로그인 API 엔드포인트 및 Basic 인증 로직 추가

- `POST /api/members/login` 매핑 구현
- `Authorization` 헤더 누락 시 `MissingAuthorizationHeaderException`(401) 발생
- 헤더가 `Basic `로 시작하지 않거나 디코딩/파싱 오류 시 `InvalidAuthorizationHeaderException`(401) 발생
- Base64 디코딩으로 이메일·비밀번호 추출
- `memberService.loginMember(email, rawPassword)` 호출하여 JWT 토큰 발급
- 잘못된 자격 증명 시 `InvalidCredentialsException`(403) 발생

5seonjae committed:
test: 회원가입 API 성공 케이스용 MemberControllerTest 추가

- `/api/members/register` 엔드포인트에 대한 통합 테스트 구현
- HTTP 201 CREATED 상태 코드 및 JSON 응답의 `token` 필드 존재 여부 검증
- `@SpringBootTest`, `@AutoConfigureMockMvc`, `@Transactional` 설정으로 MockMvc 통합 테스트 구성

5seonjae committed:
test: 회원가입 이메일 중복 시나리오 통합 테스트 추가

- 첫 번째 회원가입 요청은 HTTP 201 CREATED인지 검증
- 동일 이메일로 재가입 시 HTTP 409 CONFLICT 응답
- JSON 응답의 `error` 필드에 "이미 사용 중인 이메일입니다." 메시지 포함 여부 검사

5seonjae committed:
fix: MemberRegisterRequestDto 비밀번호 검증 어노테이션 정리

- `password` 필드에서 `@NotBlank` 제거
- `@Size(min = 6, message = "비밀번호는 최소 6자 이상이어야 합니다.")`만 적용하여 최소 길이 검증 수행

5seonjae committed:
test: 회원가입 이메일 누락 시나리오 통합 테스트 추가

- 빈 문자열 이메일(`""`)로 요청 시 HTTP 400 Bad Request 응답 검증
- JSON 응답의 `email` 필드에 "이메일은 필수입니다." 메시지 포함 여부 확인

5seonjae committed:
test: 회원가입 이메일 형식 오류 시나리오 통합 테스트 추가

- 잘못된 이메일 포맷("email")으로 요청 시 HTTP 400 Bad Request 응답 검증
- JSON 응답의 `email` 필드에 "유효한 이메일 형식이 아닙니다." 메시지 포함 여부 확인

5seonjae committed:
test: 회원가입 비밀번호 6자 미만 시나리오 통합 테스트 추가

- `password`가 6자 미만(`"123"`)인 경우 HTTP 400 Bad Request 검증
- JSON 응답의 `password` 필드에 "비밀번호는 최소 6자 이상이어야 합니다." 메시지 포함 여부 확인

5seonjae committed:
test: MemberControllerTest에 회원가입 헬퍼 메서드 `register` 추가

- `register(String email, String password)` 메서드 구현하여 테스트 내 중복 코드 제거
- 지정한 이메일과 비밀번호로 `/api/members/register` 요청 후 HTTP 201 CREATED 상태 검증

5seonjae committed:
test: 로그인 성공 케이스 통합 테스트 추가

- `login_success()` 메서드 구현
    - `register(email, password)` 헬퍼로 사전 회원가입 수행
    - Base64 인코딩된 Basic Auth 헤더를 사용해 `/api/members/login` 요청
    - HTTP 200 OK 상태 및 JSON 응답의 `token` 필드 존재 여부 검증

5seonjae committed:
test: 로그인 실패 – Authorization 헤더 누락 시나리오 테스트 추가

- `/api/members/login` 요청 시 헤더 미포함 → HTTP 401 Unauthorized 검증
- JSON 응답의 `error` 필드에 "Authorization 헤더가 필요합니다." 메시지 포함 여부 확인

5seonjae committed:
test: 로그인 실패 – 잘못된 Authorization 헤더 프리픽스 시나리오 테스트 추가

- `login_fail_invalidPrefix()` 메서드 구현
- `Authorization` 헤더에 `Bearer sometoken` 사용 시 HTTP 401 Unauthorized 및 `"Authorization 헤더 형식이 올바르지 않습니다."` 메시지 검증

5seonjae committed:
test: 로그인 실패 – 디코딩 형식 오류 시나리오 통합 테스트 추가

- 콜론 구분자 없는 Base64 문자열 사용 시 `/api/members/login` 요청
- HTTP 401 Unauthorized 상태 및 JSON 응답의 `error` 필드에 "Authorization 헤더 형식이 올바르지 않습니다." 메시지 포함 여부 검증

5seonjae committed:
refactor: 로그인 예외 처리 일관화 — InvalidCredentialsException 적용

- `MemberService.loginMember`에서 이메일·비밀번호 불일치 시 `IllegalArgumentException` 대신 `InvalidCredentialsException("이메일 또는 비밀번호가 올바르지 않습니다.")` 발생
- `MemberController.login`에서 `IllegalArgumentException` 캐치 블록 제거, 서비스에서 던진 `InvalidCredentialsException`이 직접 처리되도록 수정

5seonjae committed:
test: 로그인 실패 – 잘못된 비밀번호 시나리오 통합 테스트 추가

- `login_fail_wrongCredentials()` 메서드 구현
- HTTP 403 Forbidden 상태 및 JSON 응답의 `error` 필드에 "이메일 또는 비밀번호가 올바르지 않습니다." 메시지 검증

5seonjae committed:
docs: README 체크리스트 업데이트 및 완료 항목 표시

- 완료된 항목(회원가입 API, 로그인 API, 시나리오 테스트, API 명세)을 체크 완료 표시
- `Spring Security 및 JWT 필터 설정`과 보호된 API 구현을 선택 구현 항목으로 분리

hjjae2 commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    // 회원 생성
+    @PostMapping("/register")
```
`/register` 라는 동사가 꼭 필요할까요? 

잘 아시겠지만, RESEful API 에선 (될 수 있으면)명사 기반으로 작성하도록 권장하고 있습니다. ([참고](https://learn.microsoft.com/ko-kr/azure/architecture/best-practices/api-design#define-restful-web-api-resource-uris))

5seonjae replied:
알려주셔서 감사합니다!

REST 스타일을 의식하고 있다고 생각했는데, 여기서는 URI를 동사로 만들면 안 된다는 원칙을 깜빡했습니다. 말씀대로 클래스에 이미

@RequestMapping("/api/members")

가 선언돼 있으니, 메서드 레벨에서는 경로 없이

@PostMapping

만 남겨 두면 자연스럽게 POST /api/members 로 회원 생성이 이루어지겠네요.
곧바로 매핑·테스트·문서 경로를 모두 수정하겠습니다. 다시 한 번 세심한 피드백 감사드립니다!

hjjae2 commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -0,0 +1,72 @@
+        }
+
+        // Base64 디코딩
+        String base64Cred = authorizationHeader.substring(6).trim();
+        byte[] decoded = Base64.getDecoder().decode(base64Cred);
+        String cred = new String(decoded, StandardCharsets.UTF_8);
+        String[] parts = cred.split(":", 2);
+        if (parts.length != 2) {
+            // 디코딩 결과 형식 오류 → 401
+            throw new InvalidAuthorizationHeaderException("Authorization 헤더 형식이 올바르지 않습니다.");
+        }
```
이 처리(= Basic 헤더 파싱 처리)를 수행하는 별도의 클래스를 만들어 사용해볼까요? 역할/책임이 명확히 분리되어 관리가 용이해질 수 있을지 궁금해서요. 어떠세요?

5seonjae replied:
좋은 제안 감사합니다! 말씀하신 대로 컨트롤러가 “헤더 파싱”까지 담당하는 건 단일 책임 원칙(SRP)을 어기기 쉽습니다.
따라서 Authorization: Basic … 처리만 전담하는 클래스를 분리하려고 합니다.

- 클래스 이름 : BasicAuthHeaderParser ( 패키지 : gift.util )
- 역할
  - 헤더 형식‧Base64 디코딩 검증
  - email:password 분리
  - 결과를 레코드 Credentials(email, password) 로 반환
- 장점
  - 컨트롤러는 “요청 흐름”만 담당 → 가독성↑
  - 파서 단위 테스트 가능
  - 추후 다른 엔드포인트에서도 재사용

아래 코드는 수정한 코드입니다.
```java
// MemberController.java
    @PostMapping("/login")
    public ResponseEntity<MemberRegisterResponseDto> login(
        @RequestHeader(value = "Authorization", required = false) String authorizationHeader
    ) throws InvalidCredentialsException {

        Credentials cred = basicAuthHeaderParser.parse(authorizationHeader);

        String token = memberService.loginMember(cred.email(), cred.password());
        return ResponseEntity.ok(new MemberRegisterResponseDto(token));

    }
```

```java
package gift.util;

import gift.exception.InvalidAuthorizationHeaderException;
import gift.exception.MissingAuthorizationHeaderException;
import org.springframework.stereotype.Component;

import java.nio.charset.StandardCharsets;
import java.util.Base64;

@Component
public class BasicAuthHeaderParser {

    public Credentials parse(String authorizationHeader) {

        if (authorizationHeader == null) {
            // 헤더 누락 → 401 Unauthorized
            throw new MissingAuthorizationHeaderException("Authorization 헤더가 필요합니다.");
        }
        if (!authorizationHeader.startsWith("Basic ")) {
            // 헤더 형식 오류 → 401 Unauthorized
            throw new InvalidAuthorizationHeaderException("Authorization 헤더 형식이 올바르지 않습니다.");
        }

        // Base64 디코딩
        String base64Cred = authorizationHeader.substring(6).trim();
        byte[] decoded = Base64.getDecoder().decode(base64Cred);
        String cred = new String(decoded, StandardCharsets.UTF_8);
        String[] parts = cred.split(":", 2);
        if (parts.length != 2) {
            // 디코딩 결과 형식 오류 → 401
            throw new InvalidAuthorizationHeaderException("Authorization 헤더 형식이 올바르지 않습니다.");
        }

        String email = parts[0];
        String rawPassword = parts[1];

        return new Credentials(email, rawPassword);
    }

}
```

```java
package gift.util;

public record Credentials(String email, String password) {
}
```

hjjae2 commented on src/main/java/gift/entity/Member.java:
```diff
@@ -0,0 +1,58 @@
+        this.password = password;
+    }
+
+    // 등록 시 사용하는 생성자
```
1. "등록 시 사용하는 생성자" 라는  주석은 꼭 필요한가요? 
2.  만약 동료 개발자가 등록 외에 이 생성자가 사용한다면 어떻게 리뷰해야 할까요?

5seonjae replied:
다시 한 번 세심한 리뷰 감사드립니다! 😊
말씀 주신 포인트들을 정리해 보니, **등록 전용**이라는 주석이 오히려 코드 사용 범위를 필요 이상으로 한정시키고 있다는 점을 깨달았습니다. 제 의도는 “MemberRegisterRequestDto → Member 변환 시 아직 id 가 없다는 상황을 염두에 둔 편의 생성자” 정도였고, 반드시 등록에서만 써야 한다는 보장은 없었습니다.

### 개선 방향

- “등록 전용”이라는 표현을 빼고 일반 생성자로 두겠습니다.
필요하다면 Javadoc에 “주로 DTO → Entity 매핑 시 사용함” 정도의 _설명_만 남겨 의도를 완곡하게 전달하겠습니다.

- 비즈니스 규칙이 생기면 정적 팩토리로 분리하겠습니다.
비밀번호 암호화, 기본값 세팅 등 등록 시에만 필요한 로직이 생기면 별도 static 메서드로 캡슐화해 목적을 명확히 하겠습니다.

hjjae2 replied:
네, 좋습니다. 👍 

이해하셨겠지만, 주석은 구현을 보장해줄 순 없습니다. 보장이 필요하면 코드로 보장해줘야 합니다.

또, 주석으로 의미를 전달해주시는 것도 좋지만 의미를 잘 전달할 수 있게 네이밍에 더 신경써보세요. (주석을 반드시 지양하자 라는 의미는 아닙니다.)

hjjae2 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
에러 종류 별로 처리를 나누고 적절한 상태 코드를 잘 응답해주고 있네요. 꼼꼼한 처리 좋습니다. 👍

hjjae2 commented on src/main/java/gift/repository/MemberRepository.java:
```diff
@@ -0,0 +1,64 @@
+        return new Member(id, member.getEmail(), member.getPassword());
+    }
+
+    public boolean existByEmail(String email) {
+        String sql = "SELECT COUNT(*) FROM members WHERE email = ?";
+        Integer count = jdbcClient
+            .sql(sql)
+            .param(email)
+            .query(Integer.class)
+            .optional()
+            .orElse(0);
+
+        return count > 0;
+    }
```
https://github.com/next-step/spring-gift-wishlist/pull/82#discussion_r2180286908

이전에 위 코멘트를 주신 적이 있는데요, 이렇게 구현해보고 실행 계획(explain)을 통해 두 쿼리의 비용 차이를 확인해볼 수 있을까요?

5seonjae replied:
`COUNT(*)` vs `EXISTS` 비용 비교를 H2 메모리 DB에서 직접 측정해 보았고, 그 결과를 토대로 리포지토리 메서드 개선안을 정리했습니다.

![image](https://github.com/user-attachments/assets/da41e740-b8b0-4337-9b62-6a9de0fade6a)
![image](https://github.com/user-attachments/assets/f2dad329-7e4d-4599-b04b-f7e21742991a)

### 테스트 실험 환경

테스트는 H2 메모리 기반 DB 에서 진행했습니다.
데이터 셋은 members 100 k rows + idx_members_email(email)
H2 가 실제로 제공하는 측정 툴인 `EXPLAIN ANALYZE` 를 사용했습니다.

### 테스트 실행 계획 & 실측 결과

`SELECT COUNT(*) …`
- `execTime` : 31 ms
- `scanCount` : 2
- 집계를 위해 매치 2건 모두 스캔

`SELECT EXISTS(…)`
- `execTime` : 6 ms
- `scanCount` : 1
- 첫 행 매치 후 즉시 종료

동일 인덱스를 사용했지만 CPU 루프·스캔 횟수 차이로 약 5× 속도 차이

### 결론

- 성능 : H2 기준 5배 이상 차이가 났고, 디스크 RDBMS 환경에서는 I/O 비용까지 줄어들어 더 큰 차이가 예상됩니다.
- 가독성 : “존재 여부”를 코드·SQL 양쪽에서 표현 → 의도 명료
- 안정성 : boolean 스칼라 매핑 + single() 로 row 수 예외 처리 간결

리뷰해 주신 덕분에 실제 성능 차이를 수치로 확인하고 개선안을 적용할 수 있었습니다.
추가 피드백이나 궁금한 점이 있다면 말씀 부탁드립니다! 😊

hjjae2 replied:
실제 수치로 비교해주시니 직관적으로 이해되고 결정이 쉽네요. 👍 
어렵겠지만 앞으로도 근거에 기반한 의사 결정을 의식적으로 연습해보시죠.

hjjae2 commented on README.md:
```diff
@@ -88,4 +88,70 @@
+
+- 회원가입 (이메일/비밀번호)
+- 로그인 (Basic 인증 → JWT 발급)
+- Spring Security + JWT 필터 적용  
```
1. JWT 가 무엇인지 궁금합니다, JWT 의 특징(형식, 장/단점)에 대해 간략히 설명해주실 수 있을까요?
2. JWT 외 다른 방식으로 구현한다면 어떤 방식으로 인증(로그인 기능)을 구현해볼 수 있을까요?

5seonjae replied:
### JWT ( Json Web Token )

JWT ( Json Web Token ) 는 IETF RFC 7519 에서 정의된 토큰 포맷입니다.

Header.Payload.Signature 3 부분을 Base64URL 로 인코딩해 점 ( . ) 으로 이어 붙인 문자열로 구성되어 있습니다.

- Header : 토큰 유형 ( typ ) 과 서명 알고리즘 ( alg ) 명시
- Payload(Claims) : 사용자 식별자 ( sub ), 만료 ( exp ), 권한 ( roles ) 등 필요한 정보를 JSON 으로 담음
- Signature : 비밀 키 ( HMAC ) 또는 공개키 / 개인키 ( RSA·ECDSA ) 로 서명해 변조 방지

- 장점
  - stateless 이기 때문에, 서버가 토큰 상태를 보관하지 않아도 돼서 수평 확장에 유리
  - 자체 정보 ( Claims ) 를 포함하기 때문에 별도 조회 없이 기본 사용자·권한 정보를 얻을 수 있음
  - HTTP 헤더, WebSocket, gRPC 등 프로토콜에 구애받지 않고 전달 가능

- 단점
  - 서버가 토큰을 저장하지 않기 때문에, 탈취·로그아웃 시 exp 만료 전까지는 원칙적으로 유효함.
    즉시 폐기가 어려움
  - Payload 노출
  - 세션 ID보다 크기 때문에 네트워크·스토리지 비용이 증가할 수 있음

### JWT 외 인증 ( 로그인 ) 대안

JWT 이외에도 여러 인증 대안들이 존재합니다. 그 중 대표적인 것들만 다루어보겠습니다.

1. 세션 + 쿠키 ( stateful )

- 흐름
  - 로그인 성공 시 서버가 세션 ID 생성 → `Set-Cookie: JSESSIONID=…` 전송 → 이후 요청마다 쿠키로 식별

- 장점
  - 구현이 단순하다.
  - 서버에서 세션 객체를 삭제하기만 해도 즉시 로그아웃, 강제 종료가 가능하다.

- 단점
  - 세션 저장소가 필요하기 때문에, 수평 확장 시에 공유 비용이 발생한다.
  - 브라우저 중심의 인증 방법이다.

2. OAuth 2.0 / OpenID Connect ( SSO·소셜 로그인 )

- 흐름
  - 사용자를 외부 IdP ( Google, Kakao 등 ) 로 리다이렉트 → 사용자 승인을 거쳐 Access Token 수신

- 장점
  - 신뢰도 높은 SSO 기반·소셜 계정 재사용한다.
  - Resource Server 역할만 구현하면 되기 때문에 보안 책임을 분담할 수 있다.

- 단점
  - 초기 설정 ( 리다이렉트 URI, 클라이언트 시크릿 관리 ) 와 토큰 갱신 로직이 다소 복잡하다.

3. Basic Authentication

- 흐름
  - `Authorization: Basic Base64(id:password)` 헤더 매 요청 포함

- 장점
  - HTTP 규격에 포함된 가장 단순한 스킴
  - 테스트 API, 내부 도구용으로 빠르게 사용할 수 있다.

- 단점
  - 비밀번호가 매번 노출된다. ( HTTPS 필수 )
  - 사용자 비밀번호 유출 위험이 높다.

이외에도 Opaque Token + Introspection, API Key, Mutual TLS(mTLS), SAML 2.0 등이 있습니다.

hjjae2 replied:
> stateless 이기 때문에, 서버가 토큰 상태를 보관하지 않아도 돼서 수평 확장에 유리
> 서버가 토큰을 저장하지 않기 때문에, 탈취·로그아웃 시 exp 만료 전까지는 원칙적으로 유효함.

JWT 는 서버에 저장하면 안되나요??

5seonjae replied:
> “JWT는 서버에 저장하면 안 되나요?”

결론부터 말씀드리자면, '**저장 하면 안 된다**' 가 아니라 '**저장을 하지 않아도 된다**' 입니다.

다만 로그아웃·탈취 대응·동시 로그인 제한처럼 ‘토큰 자체를 관리해야 하는 요구’가 생기면 저장 ( = stateful 전략 ) 을 도입해도 됩니다.

이때는 “JWT + 블랙리스트 ( 또는 화이트리스트 )” 패턴을 많이 씁니다.
블랙리스트 :
- 무효화된 토큰 목록을 저장합니다. 
- 클라이언트는 블랙리스트에 등록된 토큰을 사용하려 할 때, 해당 요청이 거부됩니다. 

화이트리스트:
- 유효한 토큰 목록을 저장합니다. 
- 클라이언트는 화이트리스트에 없는 토큰을 사용하려 할 때, 해당 요청이 거부됩니다. 

### ‘저장 안 함’ 이 기본인 이유

- Stateless 설계
  - JWT 는 자체적으로 sub(사용자)·exp(만료) 등을 포함하고, 서명 검증만으로 진위 파악 가능 → DB 조회 없이도 인증 완료

- 수평 확장 용이
  - 토큰 상태를 서버에 두지 않으므로, 인스턴스 수가 늘어도 세션 복제 ( Sticky Session )·캐시 동기화 부담이 없음

- 성능
  - 매 요청마다 DB / Redis 읽지 않고, 서명만 검사 → 지연 ( latency ) ↓

따라서 관리할 필요가 없다면 “**저장하지 않는 게**” 설계·운영 면에서 더 깔끔합니다.

그래도 저장이 필요하다면, "**필요한 최소 정보**"만 보관해 Stateless 이점을 최대한 지키는 방식을 택합니다.

### 정리

원칙 : “JWT = 서버 저장 불필요” → 높은 확장성과 단순성 확보

예외 : ‘토큰을 즉시 무효화해야 할 비즈니스/보안 요구’가 있으면 부분적으로 저장해도 무방

이렇게 하면 Stateless 의 장점과 보안·운영 요건을 모두 잡을 수 있습니다.

hjjae2 replied:
크, 깔끔한 설명 감사합니다. 👍

hjjae2 commented on README.md:
```diff
@@ -88,4 +88,70 @@
+### 기능 목록
+
+- 회원가입 (이메일/비밀번호)
+- 로그인 (Basic 인증 → JWT 발급)
```
Basic 인증이 무엇인지 궁금합니다, 설명 부탁드릴게요. 😄

5seonjae replied:
### `Basic Authentication`

HTTP 헤더의 Authorization: Basic {base64(id:password)} 형식으로 사용자 자격 증명을 보내는 가장 단순한 인증 방식입니다. 자격 증명은 Base64 인코딩만 되어 있기 때문에 반드시 HTTPS 로 전송해야 안전합니다.

- 헤더 형식
  `Authorization: Basic <base64(username:password)>`

- 특징
  - Stateless - 매 요청마다 자격 증명을 전송
  - 자격 증명은 Base64 인코딩
  - 서버는 세션 저장소를 만들지 않아도 됨

- 동작 흐름
  - 클라이언트가 보호 자원 요청
    ```Plain Text
    GET /api/notes HTTP/1.1
    Host: example.com
    ```
  - 서버가 “인증 필요” 챌린지
    ```Plain Text
    HTTP/1.1 401 Unauthorized
    WWW-Authenticate: Basic realm="My API", charset="UTF-8"
    ```
  - 클라이언트가 자격 증명 생성
    1) 사용자 ID와 비밀번호를 id:password 문자열로 결합
    2) UTF-8 바이트 배열로 변환
    3) Base64 로 인코딩
  - 클라이언트가 재시도
    ```Plain Text
    GET /api/notes HTTP/1.1
    Host: example.com
    Authorization: Basic YWxhZGRpbjpvcGVuc2VzYW1l
    ```
  - 서버 인증 및 응답
    - Base64를 디코딩해서 id:password 추출
    - 사용자 DB·디렉터리·외부 IdP 등과 대조
    - 성공 시 HTTP 200 OK 와 자원 반환
    - 실패 시 HTTP 401 ( 반복 챌린지 )

- 장점
  - 구현·테스트가 매우 쉽다.
  - 서버-사이드 세션이 없어 수평 확장이 쉽다.
  - 브라우저·API 클라이언트가 기본적으로 지원한다.

- 단점
  - 자격 증명을 그대로 보내기 때문에 HTTPS 가 필수다.
  - 토큰 회수 ( Revocation )·세션 만료 같은 고급 제어가 어렵다.

---

현재 이 프로젝트는 다음과 같은 흐름을 따르고 있습니다.

- 클라이언트가 /api/members/login 에서 Basic 헤더로 이메일·패스워드 전송

- 서버가 자격 증명을 검증한 뒤 JWT 액세스 토큰을 발급

자격 증명 재전송이 필요 없고, 서버-사이드 세션도 생기지 않게 하기 위해서는 
이후 요청부터 Authorization: Bearer {JWT} 로 인증하는 기능을 구현해야 합니다.

Spring Security + OncePerRequestFilter 를 이용해서 
JWT를 해석하고 인증 컨텍스트에 넣는 필터나 인터셉터를 추가하면 위 기능을 구현할 수 있습니다.

Authorization 헤더가 Bearer로 시작하면
1) JwtParser로 서명·만료 검증
2) SecurityContextHolder에 UsernamePasswordAuthenticationToken 주입

hjjae2 replied:
> 자격 증명 재전송이 필요 없고, 서버-사이드 세션도 생기지 않게 하기 위해서는
이후 요청부터 Authorization: Bearer {JWT} 로 인증하는 기능을 구현해야 합니다.

이 부분 이해를 잘 못했는데요! Basic 인증의 경우 서버 사이드에서 세션을 저장하면 안되나요-?

5seonjae replied:
> “Basic 인증인데, 서버-사이드 세션을 만들어 쓰면 안 되나요?”

1. HTTP Basic 자체는 ‘stateless’ 인증 스킴

 Basic Auth 는 매 요청마다 `Authorization: Basic …` 헤더가 첨부되고, 그때그때 자격 증명을 검증합니다.
프로토콜 차원에서 세션 ID나 쿠키를 사용하지 않는 구조라 “stateless” 라고 합니다.

2. 세션을 만들 수도 있지만, 사실상 ‘Basic → 폼/세션 로그인’으로 모델이 바뀜

만약 서버가 Basic 헤더를 한 번만 읽고 JSESSIONID 같은 세션 쿠키를 새로 발급한다면 다음과 같은 인증 절차를 갖게 됩니다.

- 클라이언트->서버 : Basic 헤더로 1차 로그인
- 서버->클라이언트 : `Set-Cookie: JSESSIONID=…`
- 이후 요청은 쿠키 ( 세션 ) 로 인증

이렇게 되면 더 이상 Basic 헤더를 재전송하지 않으므로 “Basic Auth 기반” 이라기보다는 일반 폼/세션 로그인과 동일한 모델이 됩니다. ( 세션 유지, 서버 메모리/DB 에 Session Store 필요 )

① Basic 매 요청 
- 자격 증명 재노출 O ( 매번 ID/PW )
- 서버-사이드 세션 X
- 수평 확장 O

② Basic 1회 → 세션 쿠키 
- 자격 증명 재노출 X
- 서버-사이드 세션 O ( 세션 저장소 필요 )
- 수평 확장 - 세션 공유 필요

③ Basic 1회 → JWT ( Bearer )
- 자격 증명 재노출 X
- 서버-사이드 세션 X
- 수평 확장 O ( 토큰만 검증 )

③ 번은 ID·PW 를 한 번만 노출하면서도, 이후엔 토큰 자체에 사용자·만료 정보를 담기 때문에 세션 저장소 없이도 인증 상태를 유지할 수 있습니다.
“자격 증명 재전송이 필요 없고, 서버-사이드 세션도 생기지 않게 하려면 Bearer ( JWT ) 를 쓰자” 고 말씀드린 것입니다.

요약하면 “Basic 에서도 세션을 만들 수는 있지만, 그 순간 이미 Basic 의 stateless 특성을 버리고 세션-기반 모델로 넘어가는 셈이어서 굳이 Basic 을 사용할 이유가 줄어든다”는 점을 강조하고 싶었습니다. 🙂 
필요한 부분 더 궁금하시면 언제든 말씀 주세요!

hjjae2 replied:
어우, 잘 정리해주셔서 감사합니다. 덕분에 저도 잘 이해가 됐네요. 👍 

사소한데요, 2, 3 안에 `자격 증명 재노출 X` 라고 표현해주신 건 조금 더 정확히 표현할 필요가 있을 것 같아요. 아마 말씀주신 건 ID/PW 이겠죠?

관점에 따라 "자격 증명"이 표현하는게 토큰이 될 수도, ID/PW가 될 수도 있기에 말씀드립니다. 🙇 

상세한 설명 감사합니다. 👏

hjjae2 commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -0,0 +1,72 @@
+
+@RestController
+@RequestMapping("/api/members")
+public class MemberController {
```
강의 시간에 잠깐 배우셨을거에요. 이렇게 작성해주신 컨트롤러는 `DispatcherServlet` 라는 녀석의 도움을 받아 사용자의 요청과 매핑되어 수행되게 되는데요. `DispatcherServlet` 에 대해 간략히 설명을 부탁드려도 될까요? 🙏

5seonjae replied:
### `DispatcherServlet` 의  정의

DispatcherServlet은 Spring MVC가 Front Controller 패턴을 구현한 핵심 서블릿입니다. 모든 HTTP 요청을 가장 먼저 받아 내부의 전략(Strategy) 빈에게 위임함으로써, 컨트롤러·뷰 해석·예외 처리 흐름을 표준화합니다.

### `DispatcherServlet` 의 역할

`DispatcherServlet` 의 역할은 크게 3가지가 있습니다.

1. Front Controller

  하나의 서블릿이 모든 HTTP 요청을 받아서 내부 컴포넌트에게 위임하는 패턴을 구현합니다.

2. 요청–응답 파이프라인 오케스트레이션

`HandlerMapping`, `HandlerAdapter`, `ViewResolver`, `HandlerExceptionResolver` 등을 순서대로 호출해 전체 흐름을 제어합니다.

전체적인 흐름은 다음과 같습니다.

- `Servlet Container` 가 요청을 수신하고, `Servlet Filter` 들이 호출됩니다.

- 이후에 URL 패턴과 매핑된 `DispatcherServlet` 이 가장 먼저 요청을 받습니다.

- `HandlerMapping` 목록을 순회해서 URL, HTTP Method 에 맞는 컨트롤러 ( 핸들러 ) 와 해당 요청에 적용될 `HandlerInterceptor` 목록을 묶어 `HandlerExecutionChain` 을 생성합니다.

- 체인에 포함된 `HandlerInterceptor` 들의 `preHandle()` 메서드가 등록 순서대로 호출합니다.

- `HandlerAdapter` 가 선택된 컨트롤러 메서드를 호출합니다.

- Controller 결과를 처리합니다.

- 컨트롤러가 정상 종료되면 postHandle() ( preHandle() 의 역순 ) → 뷰 렌더링 전에 모델을 수정할 기회를 얻습니다.
  만약 Controller, Interceptor 에서 예외가 발생하면 `HandlerExceptionResolver` 들이 순서대로 예외를 HTTP 응답으로 변환합니다. ( 예시 : `@ExceptionHandler`, `ResponseStatusExceptionResolver` )

- ViewResolver가 선택한 View 객체가 render() 를 호출해 HTML 등을 작성하거나,
HttpMessageConverter가 객체를 JSON 바디로 직렬화해 HttpServletResponse 에 기록합니다.

- 응답 본문이 완전히 작성된 뒤, afterCompletion() ( 모든 인터셉터의 역순 ) 호출 → 로깅·리소스 해제·트랜잭션 종료
  DispatcherServlet 이 제어를 서블릿 컨테이너로 넘기고, 컨테이너가 HTTP 응답을 클라이언트로 전송합니다.

3. Spring IoC 통합

위임 대상 객체 ( 컨트롤러·뷰 리졸버 등 ) 는 Spring 컨테이너에서 빈으로 주입받으므로, DI·AOP 같은 Spring 기능을 그대로 활용할 수 있습니다.

### `DispatcherServlet` 의 장점

1. 중앙 집중 제어 형식으로 인해 모든 HTTP 요청이 한 곳으로 들어오기 때문에 라우팅, 보안, 모니터링 같은 전역적인 정책을 한 층에서 관리할 수 있다.

2. `HandlerMapping`, `ViewResolver`, `HandlerExceptionResolver` 등을 원하는 구현체로 손쉽게 바꿀 수 있어서 JSP·Thymeleaf·JSON·GraphQL 등 다양한 워크플로를 동일 파이프라인에서 지원한다.

3. 컨트롤러·인터셉터·리졸버가 전부 Bean 이기 때문에 DI, AOP, 프로퍼티 설정을 그대로 활용할 수 있다.

4. Filter, HandlerInterceptor 와 조합해서 인증, 로깅, 트랜잭션을 깔끔하게 분리할 수 있다.

5. 핸들러 메서드 단위로 모킹, 슬라이스 테스트가 가능하기 때문에 빠른 단위 테스트를 작성할 수 있다.

6. Spring Boot 자동 구성으로 빠르게 개발할 수 있다.

7. 일관된 예외 처리와 국제화를 지원한다.

### `DispatcherServlet` 의 단점

1. 중앙 집중 제어 형식으로 인해 모든 HTTP 요청이 한 곳으로 들어오기 때문에 잘못된 설정이나 Filter-Interceptor 무한 루프가 발생하면 모든 요청이 실패한다.

2. 다층 추상화로 인해 내부 작동 방식에 대해 이해하기 힘들다.

3. 요청 : 쓰레드 = 1 : 1 방식이기 때문에, 대량 동시 접속 시에는 쓰레드, 메모리를 많이 소모한다.

4. 자동 설정이 오히려 디버깅을 어렵게 만들 수 있다.

5. Reflection, Message Converter, 다중 Interceptor Chaining 으로 지연이 다소 늘어난다.

hjjae2 commented on src/main/java/gift/service/TokenService.java:
```diff
@@ -0,0 +1,35 @@
+public class TokenService {
+
+    private final Key key;
+    private final long expirationMs;
```
`expirationMs` 값은 어느 정도로 설정하실 계획이신지 궁금합니다-! 🙇

5seonjae replied:
만료 시간(expiration-ms)을 3 600 000 ms(1 시간)으로 설정했습니다.

- JWT 특성상 탈취 위험을 줄이려면 TTL을 너무 길게 잡지 않는 것이 좋고,
- 반대로 너무 짧으면 사용자가 자주 재인증해야 해 UX가 나빠집니다.

그래서 실무 가이드에서 흔히 권장하는 15 분 ~ 1 시간 범위 중
**1 시간**을 우선 채택했습니다.  
운영 지표(세션 길이·로그인 빈도)를 보고 필요하면
30 분이나 2 시간 등으로 조정할 수 있도록
`jwt.expiration-ms`를 환경 변수로 외부화해 두었습니다.

hjjae2 replied:
> 운영 지표(세션 길이·로그인 빈도)를 보고 필요하면 30 분이나 2 시간 등으로 조정할 수 있도록

현실적인 운영 방식이네요! 👍👍

hjjae2 commented on .gitignore:
```diff
@@ -38,3 +38,6 @@ out/
 .DS_Store
+
+# 환경 변수 파일
+.env
```
`.env` 파일은 왜 ignore 처리가 된건가요-? 🙇

hjjae2 reviewed (changes requested):
안녕하세요, 선재님! 이번 단계도 꼼꼼하게 잘 구현해주셨네요, 고생하셨습니다. 👍 

몇 가지 코멘트 남겼는데요, 확인 부탁드릴게요. 💪

5seonjae committed:
refactor: 회원가입 API 경로 기본 매핑으로 변경 및 테스트 수정

- `MemberController`의 `@PostMapping("/register")`을 `@PostMapping()`으로 변경
- `MemberControllerTest` 내 회원가입 관련 테스트들의 요청 URL을 새로운 매핑에 맞게 업데이트

5seonjae committed:
refactor: Member, Product 클래스 생성자 관련 주석 제거

- `// 등록 시 사용하는 생성자` 주석을 삭제하여 생성자가 특정 용도에 국한되지 않음을 명확히 함

5seonjae committed:
refactor: BasicAuthHeaderParser 도입 및 로그인 컨트롤러 헤더 파싱 로직 분리

- `BasicAuthHeaderParser` 컴포넌트 추가하여 `Authorization` 헤더 파싱 책임 분리
- `Credentials` 레코드로 이메일·비밀번호를 캡슐화
- `MemberController.login`에서 기존 파싱 로직 제거 후 `BasicAuthHeaderParser.parse` 호출 적용

5seonjae committed:
refactor: isApproved 메서드 쿼리를 COUNT → EXISTS로 최적화

- `SELECT COUNT(*) ...` 구문을 `SELECT EXISTS (SELECT 1 ...)` 서브쿼리로 변경
- `jdbcClient.query(Boolean.class).single()` 사용해 Boolean 결과 직접 반환

5seonjae committed:
refactor: MemberRepository.existByEmail 쿼리 COUNT → EXISTS로 최적화

- `SELECT COUNT(*) …` 대신 `SELECT EXISTS (SELECT 1 …)` 서브쿼리 사용
- `jdbcClient.query(Boolean.class).single()`로 Boolean 결과 직접 반환

hjjae2 reviewed (approved):
안녕하세요, 선재님!

코멘트 드린 부분에 대해 상세히 잘 확인해주셔서 감사합니다. 🙇 잘 확인해주신 덕에 이번 단계는 여기서 마무리해도 좋겠습니다. 머지하고 다음 단계에서 뵙겠습니다.

이와 별개로 추가로 궁금한 내용을 코멘트 남겨두었는데요. 이것들은 확인 부탁드릴게요. 🙏

hjjae2 commented:
> 다음 단계에서 Spring Security 및 JWT 필터 적용, 보호된 API 엔드포인트 구현 예정

참고해주세요. 다음 단계 구현 시, Spring Security 는 사용하지 않고 구현해주시기 바랍니다. 

미션의 요구사항은 Spring Security 없이도 충분히 구현 가능하며, 불필요한 복잡도는 제거하는 것이 목적입니다. 🙇
